traversal/scanner/GraphIterator.java

```diff
@@ -176,82 +214,29 @@ private boolean computeAnswer() {
-        Set<ProcedureVertex<?, ?>> verifyFailureCauses = new HashSet<>();
-        while (!verified && vertexTraverser.hasNextVertex()) {
-            vertexTraverser.takeNextVertex();
-            if (vertexTraverser.verifyLoops()
-                    && verifyTraversedScopes(procedureVertex, verifyFailureCauses)
-                    && verifyEdgeLookahead(procedureVertex, verifyFailureCauses)) {
-                verified = true;
-            } else {
-                vertexTraverser.clearCurrentVertex();
-                removeFromSuccessors(procedureVertex);
-            }
-        }
-        toRevisit.addAll(verifyFailureCauses);
-        if (verified) procedureVertex.outs().forEach(edge -> toTraverse.add(edge.to()));
-        else failed(procedureVertex);
+        if (vertexTraverser.findNextVertex()) success(vertexTraverser);
+        else failed(vertexTraverser);
     }
 
-    private void removeFromSuccessors(ProcedureVertex<?, ?> procedureVertex) {
-        procedureVertex.outs().forEach(edge -> {
-            vertexTraversers.get(edge.to()).removeEdgeInput(edge);
-            removeFromSuccessors(edge.to());
-        });
+    private void success(VertexTraverser vertexTraverser) {
+        iterate(vertexTraverser.procedureVertex.dependents()).link(iterate(vertexTraverser.implicitDependents))
+                .forEachRemaining(v -> {
+                    toTraverse.add(v);
+                    vertexTraversers.get(v).clear();
+                });
     }
 
-    private void failed(ProcedureVertex<?, ?> procedureVertex) {
-        procedureVertex.ins().forEach(edge -> toRevisit.add(edge.from()));
-        vertexTraversers.get(procedureVertex).reset();
-        toTraverse.add(procedureVertex);
+    private void failed(VertexTraverser vertexTraverser) {
+        iterate(vertexTraverser.procedureVertex.dependees()).forEachRemaining(dependee -> toRevisit.add(dependee));
+        vertexTraverser.implicitDependees.forEach(toRevisit::add);
+        vertexTraverser.clear();
+        toTraverse.add(vertexTraverser.procedureVertex);
         direction = Direction.REVISIT;
     }
 
-    private boolean verifyTraversedScopes(ProcedureVertex<?, ?> procedureVertex, Set<ProcedureVertex<?, ?>> verifyFailureCauses) {
-        if (procedureVertex.id().isScoped()) {
-            Scopes.Scoped scoped = scopes.get(procedureVertex.id().asScoped().scope());
-            return verifyScopedUpToOrder(scoped, procedureVertex.order(), verifyFailureCauses);
-        } else if (procedureVertex.isThing() && !procedureVertex.asThing().inRolePlayers().isEmpty()) {
-            for (ProcedureEdge<?, ?> edge : procedureVertex.asThing().inRolePlayers()) {
-                assert edge.isRolePlayer();
-                Scopes.Scoped scoped = scopes.get(edge.asRolePlayer().scope());
-                if (!verifyScopedUpToOrder(scoped, procedureVertex.order(), verifyFailureCauses)) return false;
-            }
-        }
-        return true;
-    }
```

Comment:
scopes are now purely handled by filtering & vertex backtracking. The latter handles all complex cases when considered with the fact that we guarantee de-optimisation of partially overlapping role traversals